aws_codedeploy_deployment_group on_premises_instance_tag_filter not added after apply

I have a simple deployment group defined as
```
resource "aws_codedeploy_deployment_group" "qa" {
  app_name = "${aws_codedeploy_app.my_app.name}"
  deployment_group_name = "qa"
  service_role_arn      = "${terraform_remote_state.codedeploy.output.aws_iam_role_codedeploy_service_role_arn}"
  deployment_config_name = "CodeDeployDefault.OneAtATime"
  on_premises_instance_tag_filter {
    type = "VALUE_ONLY"
    value = "staging-qa"
  }
}
```

When I run terraform plan, I can see that the deployment group will be created
````
+ aws_codedeploy_deployment_group.qa
    app_name:                                         "" => "My App"
    deployment_config_name:                           "" => "CodeDeployDefault.OneAtATime"
    deployment_group_name:                            "" => "qa"
    on_premises_instance_tag_filter.#:                "" => "1"
    on_premises_instance_tag_filter.1446378810.key:   "" => ""
    on_premises_instance_tag_filter.1446378810.type:  "" => "VALUE_ONLY"
    on_premises_instance_tag_filter.1446378810.value: "" => "staging-qa"
    service_role_arn:                                 "" => "arn:aws:iam::123456789012:role/CodeDeployServiceRole"
````

After running terraform apply with no error, when I go into AWS to verify the deployment group, the on premise instance tag is not added.

@stack72 second apply added the tag for me too, but crash terraform and corrupt state file.
```
panic: runtime error: invalid memory address or nil pointer dereference
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws: [signal 0xb code=0x1 addr=0x8 pc=0x4ffa5a]
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws:
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws: goroutine 133 [running]:
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws: panic(0x14027e0, 0xc82000a0f0)
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws:     /opt/go/src/runtime/panic.go:481 +0x3e6
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws: github.com/hashicorp/terraform/builtin/providers/aws.onPremisesTagFiltersToMap(0xc8200297e0, 0x1, 0x1, 0x0, 0x0, 0x0)
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws:     /opt/gopath/src/github.com/hashicorp/terraform/builtin/providers/aws/resource_aws_codedeploy_deployment_group.go:403 +0x11a
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws: github.com/hashicorp/terraform/builtin/providers/aws.resourceAwsCodeDeployDeploymentGroupRead(0xc820384fc0, 0x10970e0, 0xc820396000, 0x0, 0x0)
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws:     /opt/gopath/src/github.com/hashicorp/terraform/builtin/providers/aws/resource_aws_codedeploy_deployment_group.go:252 +0x761
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws: github.com/hashicorp/terraform/builtin/providers/aws.resourceAwsCodeDeployDeploymentGroupUpdate(0xc820384fc0, 0x10970e0, 0xc820396000, 0x0, 0x0)
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws:     /opt/gopath/src/github.com/hashicorp/terraform/builtin/providers/aws/resource_aws_codedeploy_deployment_group.go:306 +0xa90
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws: github.com/hashicorp/terraform/helper/schema.(*Resource).Apply(0xc8204248c0, 0xc8201dec00, 0xc8204624e0, 0x10970e0, 0xc820396000, 0xc820108101, 0x0, 0x0)
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws:     /opt/gopath/src/github.com/hashicorp/terraform/helper/schema/resource.go:152 +0x3b9
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws: github.com/hashicorp/terraform/helper/schema.(*Provider).Apply(0xc82048ba40, 0xc820310d80, 0xc8201dec00, 0xc8204624e0, 0x227f501, 0x0, 0x0)
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws:     /opt/gopath/src/github.com/hashicorp/terraform/helper/schema/provider.go:162 +0x1ed
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws: github.com/hashicorp/terraform/rpc.(*ResourceProviderServer).Apply(0xc8204ca420, 0xc820488920, 0xc820462580, 0x0, 0x0)
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws:     /opt/gopath/src/github.com/hashicorp/terraform/rpc/resource_provider.go:323 +0x76
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws: reflect.Value.call(0x133bf20, 0x16481d8, 0x13, 0x16fff20, 0x4, 0xc82035bed8, 0x3, 0x3, 0x0, 0x0, ...)
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws:     /opt/go/src/reflect/value.go:435 +0x120d
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws: reflect.Value.Call(0x133bf20, 0x16481d8, 0x13, 0xc82035bed8, 0x3, 0x3, 0x0, 0x0, 0x0)
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws:     /opt/go/src/reflect/value.go:303 +0xb1
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws: net/rpc.(*service).call(0xc82033cc40, 0xc82033cc00, 0xc8204c5620, 0xc82041c400, 0xc82048d180, 0x1097f80, 0xc820488920, 0x16, 0x1097fe0, 0xc820462580, ...)
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws:     /opt/go/src/net/rpc/server.go:383 +0x1c2
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws: created by net/rpc.(*Server).ServeCodec
2016/05/11 15:57:10 [DEBUG] terraform-provider-aws:     /opt/go/src/net/rpc/server.go:477 +0x49d
```

--- a/builtin/providers/aws/resource_aws_codedeploy_deployment_group.go
+++ b/builtin/providers/aws/resource_aws_codedeploy_deployment_group.go
@@ -177,7 +177,7 @@ func resourceAwsCodeDeployDeploymentGroupCreate(d *schema.ResourceData, meta int
 	if attr, ok := d.GetOk("autoscaling_groups"); ok {
 		input.AutoScalingGroups = expandStringList(attr.(*schema.Set).List())
 	}
-	if attr, ok := d.GetOk("on_premises_instance_tag_filters"); ok {
+	if attr, ok := d.GetOk("on_premises_instance_tag_filter"); ok {
 		onPremFilters := buildOnPremTagFilters(attr.(*schema.Set).List())
 		input.OnPremisesInstanceTagFilters = onPremFilters
 	}
@@ -331,9 +331,15 @@ func buildOnPremTagFilters(configured []interface{}) []*codedeploy.TagFilter {
 		var filter codedeploy.TagFilter
 		m := raw.(map[string]interface{})
 
-		filter.Key = aws.String(m["key"].(string))
-		filter.Type = aws.String(m["type"].(string))
-		filter.Value = aws.String(m["value"].(string))
+		if v, ok := m["key"]; ok {
+			filter.Key = aws.String(v.(string))
+		}
+		if v, ok := m["type"]; ok {
+			filter.Type = aws.String(v.(string))
+		}
+		if v, ok := m["value"]; ok {
+			filter.Value = aws.String(v.(string))
+		}
 
 		filters = append(filters, &filter)
 	}
@@ -400,13 +406,13 @@ func onPremisesTagFiltersToMap(list []*codedeploy.TagFilter) []map[string]string
 	result := make([]map[string]string, 0, len(list))
 	for _, tf := range list {
 		l := make(map[string]string)
-		if *tf.Key != "" {
+		if tf.Key != nil && *tf.Key != "" {
 			l["key"] = *tf.Key
 		}
-		if *tf.Value != "" {
+		if tf.Value != nil && *tf.Value != "" {
 			l["value"] = *tf.Value
 		}
-		if *tf.Type != "" {
+		if tf.Type != nil && *tf.Type != "" {
 			l["type"] = *tf.Type
 		}
 		result = append(result, l)

--- a/builtin/providers/aws/resource_aws_codedeploy_deployment_group_test.go
+++ b/builtin/providers/aws/resource_aws_codedeploy_deployment_group_test.go
@@ -69,6 +69,39 @@ func TestAccAWSCodeDeployDeploymentGroup_basic(t *testing.T) {
 	})
 }
 
+func TestAccAWSCodeDeployDeploymentGroup_onPremiseTag(t *testing.T) {
+	var group codedeploy.DeploymentGroupInfo
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSCodeDeployDeploymentGroupDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccAWSCodeDeployDeploymentGroupOnPremiseTags,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSCodeDeployDeploymentGroupExists("aws_codedeploy_deployment_group.foo", &group),
+					resource.TestCheckResourceAttr(
+						"aws_codedeploy_deployment_group.foo", "app_name", "foo_app"),
+					resource.TestCheckResourceAttr(
+						"aws_codedeploy_deployment_group.foo", "deployment_group_name", "foo"),
+					resource.TestCheckResourceAttr(
+						"aws_codedeploy_deployment_group.foo", "deployment_config_name", "CodeDeployDefault.OneAtATime"),
+
+					resource.TestCheckResourceAttr(
+						"aws_codedeploy_deployment_group.foo", "on_premises_instance_tag_filter.#", "1"),
+					resource.TestCheckResourceAttr(
+						"aws_codedeploy_deployment_group.foo", "on_premises_instance_tag_filter.2916377465.key", "filterkey"),
+					resource.TestCheckResourceAttr(
+						"aws_codedeploy_deployment_group.foo", "on_premises_instance_tag_filter.2916377465.type", "KEY_AND_VALUE"),
+					resource.TestCheckResourceAttr(
+						"aws_codedeploy_deployment_group.foo", "on_premises_instance_tag_filter.2916377465.value", "filtervalue"),
+				),
+			},
+		},
+	})
+}
+
 func TestAccAWSCodeDeployDeploymentGroup_triggerConfiguration_basic(t *testing.T) {
 	var group codedeploy.DeploymentGroupInfo
 
@@ -516,6 +549,71 @@ resource "aws_codedeploy_deployment_group" "foo" {
 		key = "filterkey"
 		type = "KEY_AND_VALUE"
 		value = "anotherfiltervalue"
+	}
+}`
+
+var testAccAWSCodeDeployDeploymentGroupOnPremiseTags = `
+resource "aws_codedeploy_app" "foo_app" {
+	name = "foo_app"
+}
+
+resource "aws_iam_role_policy" "foo_policy" {
+	name = "foo_policy"
+	role = "${aws_iam_role.foo_role.id}"
+	policy = <<EOF
+{
+	"Version": "2012-10-17",
+	"Statement": [
+		{
+			"Effect": "Allow",
+			"Action": [
+				"autoscaling:CompleteLifecycleAction",
+				"autoscaling:DeleteLifecycleHook",
+				"autoscaling:DescribeAutoScalingGroups",
+				"autoscaling:DescribeLifecycleHooks",
+				"autoscaling:PutLifecycleHook",
+				"autoscaling:RecordLifecycleActionHeartbeat",
+				"ec2:DescribeInstances",
+				"ec2:DescribeInstanceStatus",
+				"tag:GetTags",
+				"tag:GetResources"
+			],
+			"Resource": "*"
+		}
+	]
+}
+EOF
+}
+
+resource "aws_iam_role" "foo_role" {
+	name = "foo_role"
+	assume_role_policy = <<EOF
+{
+	"Version": "2012-10-17",
+	"Statement": [
+		{
+			"Sid": "",
+			"Effect": "Allow",
+			"Principal": {
+				"Service": [
+					"codedeploy.amazonaws.com"
+				]
+			},
+			"Action": "sts:AssumeRole"
+		}
+	]
+}
+EOF
+}
+
+resource "aws_codedeploy_deployment_group" "foo" {
+	app_name = "${aws_codedeploy_app.foo_app.name}"
+	deployment_group_name = "foo"
+	service_role_arn = "${aws_iam_role.foo_role.arn}"
+	on_premises_instance_tag_filter {
+		key = "filterkey"
+		type = "KEY_AND_VALUE"
+		value = "filtervalue"
 	}
 }`
 

